### PR TITLE
fix: [PIE-8919]: RBAC validation in secret retrieval in the pipeline execution steps

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes

**Issue:** A user without having access permission for a secret is able to refer to that in a step.

**Changes:** Add RBAC check inside the Secret functor while resolving the value of the secret.

**Impact:** Users won't be access to refer to a secret for which they don't have access permissions.

**Testing:** Tested the following cases:
1. Executing a pipeline with triggers comes with elevated principals and they can decrypt any secrets. 
2. Using a secret in a variable and then refer that variable in a step to decrypt the value of secret. An error will be thrown in this case, if a user doesn't have access permission on the referred secret.
3. Using account, org, and project-level secrets in a pipeline, the following error is shown when a secret without having its access is referred to in a step.
4. Error thrown in case of empty secret identifier

Screenshot 2023-03-20 at 5 53 52 PM

Screenshot 2023-03-22 at 10 37 01 AM

Screenshot 2023-03-27 at 1 58 16 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)

### CODE REVIEW

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 1]
```diff
@@ -656,7 +656,8 @@ public enum FeatureName {
   NG_CDS_NATIVE_EKS_SUPPORT("Enable native EKS support for K8s/Native Helm Infrastructures", HarnessTeam.CDP),
   CDS_RANCHER_SUPPORT_NG("Enable Rancher support in NG.", HarnessTeam.CDP),
   PL_SELECT_SPECIFIC_SERVICE_ACCOUNT_IN_RESOURCE_GROUP(
-      "Allow selection of specific service accounts in Resource group", HarnessTeam.PL);
+      "Allow selection of specific service accounts in Resource group", HarnessTeam.PL),
+  USE_SECRET_FUNCTOR_WITH_RBAC("Perform Rbac on secrets when used in pipeline execution", HarnessTeam.PIPELINE);
```
Comment (naidu-harness) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:660]: FF name should start with our project in this case PIE

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 36 at r4](https://reviewable.io/reviews/harness/harness-core/45343#-NR1MYRO4vI7JK7wh5yg:-NR1MYRO4vI7JK7wh5yh:b-1fvvdq) ([raw file](https://github.com/harness/harness-core/blob/a2d350734d1d61b5a8f95c4114179a865eccd260/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L36)):*
> ```Java
>       IdentifierRefProtoDTOHelper identifierRefProtoDTOHelper) {
>     this.ambiance = ambiance;
>     this.expressionFunctorToken = ambiance.getExpressionFunctorToken();
> ```

nit: Is this variable required?

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java` line 190 at r4](https://reviewable.io/reviews/harness/harness-core/45343#-NR1PyN5ECdeEFRzq2Ml:-NR1PyN5ECdeEFRzq2Mm:byy63ym) ([raw file](https://github.com/harness/harness-core/blob/a2d350734d1d61b5a8f95c4114179a865eccd260/pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java#L190)):*
> ```Java
>     });
>     ServiceHttpClientConfig serviceHttpClientConfig =
>         ServiceHttpClientConfig.builder().baseUrl("http://localhost:7457/").build();
> ```

I think this should come from config?

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java` line 190 at r4](https://reviewable.io/reviews/harness/harness-core/45343#-NR1PyN5ECdeEFRzq2Ml:-NR1Q6FI2GSf9A5TM2BZ:bcx2rlh) ([raw file](https://github.com/harness/harness-core/blob/a2d350734d1d61b5a8f95c4114179a865eccd260/pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java#L190)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

I think this should come from config?
</blockquote></details>

nvm, it is test file

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 36 at r4](https://reviewable.io/reviews/harness/harness-core/45343#-NR1MYRO4vI7JK7wh5yg:-NR1R--D4rqRs_w5jPfz:b-xhwoaf) ([raw file](https://github.com/harness/harness-core/blob/a2d350734d1d61b5a8f95c4114179a865eccd260/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L36)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

nit: Is this variable required?
</blockquote></details>

not really, it's only being used in one place, we can directly pass its value from the ambiance.

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 1 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 4 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by naidu-harness — DISMISSED** [review]
> Reviewed 1 of 2 files at r2, 1 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 1. unresolved discussions.

**Review by hextechpal — DISMISSED** [review]
> Reviewed 1 of 2 files at r2, 1 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 1. unresolved discussions.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r2, 1 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 30 at r6](https://reviewable.io/reviews/harness/harness-core/45343#-NR2ie2x00p4U5L4JyLy:-NR2ie2x00p4U5L4JyLz:b-2u3sit) ([raw file](https://github.com/harness/harness-core/blob/d320247e02de3a04a2479f7d8c932d0b9542ad61/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L30)):*
> ```Java
>   PipelineRbacHelper pipelineRbacHelper;
>   Ambiance ambiance;
>   IdentifierRefProtoDTOHelper identifierRefProtoDTOHelper;
> ```

This helper should be a util class -> IdentifierRefProtoDTOHelper meaning all methods static, can we please make it, then you will not need to inject it



_Code quote:_
```Java
createIdentifierRefProtoDTO
```
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 47 at r6](https://reviewable.io/reviews/harness/harness-core/45343#-NR2jaHs1aRHqR91Naoe:-NR2jaHs1aRHqR91Naof:b4t91n0) ([raw file](https://github.com/harness/harness-core/blob/d320247e02de3a04a2479f7d8c932d0b9542ad61/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L47)):*
> ```Java
>                      .setIdentifierRef(identifierRefProtoDTOHelper.createIdentifierRefProtoDTO(
>                          AmbianceUtils.getAccountId(ambiance), AmbianceUtils.getOrgIdentifier(ambiance),
>                          AmbianceUtils.getProjectIdentifier(ambiance), secretIdentifier))
> ```

We have 3 types of secrets, account, org and project here we are always assuming project identifier.

IdentifierRefHelper.getIdentifierRef helps you get scoped ref. You can check code and retest and add details to Testing in pr as well



_Code quote:_
```Java
                     .setIdentifierRef(identifierRefProtoDTOHelper.createIdentifierRefProtoDTO(
                         AmbianceUtils.getAccountId(ambiance), AmbianceUtils.getOrgIdentifier(ambiance),
                         AmbianceUtils.getProjectIdentifier(ambiance), secretIdentifier))
```

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 3 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 47 at r6](https://reviewable.io/reviews/harness/harness-core/45343#-NR2jaHs1aRHqR91Naoe:-NR3czwy5FxiDcKu0vk7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d320247e02de3a04a2479f7d8c932d0b9542ad61/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L47)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

We have 3 types of secrets, account, org and project here we are always assuming project identifier.

IdentifierRefHelper.getIdentifierRef helps you get scoped ref. You can check code and retest and add details to Testing in pr as well
</blockquote></details>

Done.

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 4 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 30 at r6](https://reviewable.io/reviews/harness/harness-core/45343#-NR2ie2x00p4U5L4JyLy:-NR6T1io0HqJVj0sOP48:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d320247e02de3a04a2479f7d8c932d0b9542ad61/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L30)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

This helper should be a util class -> IdentifierRefProtoDTOHelper meaning all methods static, can we please make it, then you will not need to inject it
</blockquote></details>

Done.

**Comment by shalini-agr** [comment]
trigger messagecheck

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r8, 11 of 11 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 14 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 52 at r9](https://reviewable.io/reviews/harness/harness-core/45343#-NRBmvDbEhevP3ULRoaQ:-NRBmvDc31vhbOUaOF_6:b-7oheqx) ([raw file](https://github.com/harness/harness-core/blob/81d123dddd3ff9934a532311e80afdf2a417ded8/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L52)):*
> ```Java
>       return "${ngSecretManager.obtain(\"" + secretIdentifier + "\", " + ambiance.getExpressionFunctorToken() + ")}";
>     } catch (NGAccessDeniedException exception) {
>       throw new EngineFunctorException(exception.getMessage(), exception);
> ```

why we are throing functor exception? shouldnt this we access denied exception only? can you share exception in API which is thrown?

**Review by archit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 14 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 52 at r9](https://reviewable.io/reviews/harness/harness-core/45343#-NRBmvDbEhevP3ULRoaQ:-NRBp02h4qiS7gfbQW4R:b6an6jr) ([raw file](https://github.com/harness/harness-core/blob/81d123dddd3ff9934a532311e80afdf2a417ded8/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L52)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why we are throing functor exception? shouldnt this we access denied exception only? can you share exception in API which is thrown?
</blockquote></details>

done, discussed offline

**Review by sahilhindwani-harness — DISMISSED** [review]
> Reviewed 3 of 4 files at r8, 10 of 11 files at r9, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 1 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 1. unresolved discussions.

**Comment by mohitg0795** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 51 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHbG34wXKPupZIrDB:-NRHHbG34wXKPupZIrDC:b-98jaci) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L51)):*
> ```Java
>       pipelineRbacHelper.checkRuntimePermissions(ambiance, entityDetails);
>       return "${ngSecretManager.obtain(\"" + secretIdentifier + "\", " + ambiance.getExpressionFunctorToken() + ")}";
>     } catch (NGAccessDeniedException exception) {
> ```

we should log exception here? Do we know if this exception is going to be centrally logged? If not, then its a must to log it here.

**Comment by mohitg0795** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 50 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHhvNDmIO9nWQh1yX:-NRHHhvNDmIO9nWQh1yY:b-ximdyn) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L50)):*
> ```Java
>                      .build());
>       pipelineRbacHelper.checkRuntimePermissions(ambiance, entityDetails);
>       return "${ngSecretManager.obtain(\"" + secretIdentifier + "\", " + ambiance.getExpressionFunctorToken() + ")}";
> ```

This looks too much hard-coding. Can't we use such expressions from a common utility class?

**Comment by mohitg0795** [comment]
*[`pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java` line 193 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHrf71-iinUJmW7YV:-NRHHrf71-iinUJmW7YW:b-86bz0l) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java#L193)):*
> ```Java
>     modules.add(AccessControlClientModule.getInstance(AccessControlClientConfiguration.builder()
>                                                           .enableAccessControl(true)
>                                                           .accessControlServiceSecret("secret")
> ```

please use constants

**Comment by mohitg0795** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/AmbianceExpressionEvaluator.java` line 122 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHI-R2BUcYlojLZWqG:-NRHI-R2BUcYlojLZWqH:b-n6db23) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/AmbianceExpressionEvaluator.java#L122)):*
> ```Java
>       if (pmsFeatureFlagService.isEnabled(
>               AmbianceUtils.getAccountId(ambiance), FeatureName.PIE_USE_SECRET_FUNCTOR_WITH_RBAC)) {
>         addToContext("secrets", new SecretFunctorWithRbac(ambiance, pipelineRbacHelper));
> ```

please use constants

**Comment by mohitg0795** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 53 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHIKP8DbK4FDg3iQnj:-NRHIKP8DbK4FDg3iQnk:b-mf4i5e) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L53)):*
> ```Java
>     } catch (NGAccessDeniedException exception) {
>       throw new EngineFunctorException(exception.getMessage(), exception);
>     }
> ```

what happens if we get any generic exception? Is it fine to not catch it here?

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 6. unresolved discussions. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/AmbianceExpressionEvaluator.java` line 122 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHI-R2BUcYlojLZWqG:-NRHpU2C6kZCSUPfFDKj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/AmbianceExpressionEvaluator.java#L122)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

please use constants
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 50 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHhvNDmIO9nWQh1yX:-NRHrdum86Wt-PMtLIj3:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L50)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This looks too much hard-coding. Can't we use such expressions from a common utility class?
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 51 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHbG34wXKPupZIrDB:-NRHpW5m0NuoqwF6MHT3:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L51)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

we should log exception here? Do we know if this exception is going to be centrally logged? If not, then its a must to log it here.
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 53 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHIKP8DbK4FDg3iQnj:-NRHfXQH50KUVgTK43TX:b-2vteae) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L53)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

what happens if we get any generic exception? Is it fine to not catch it here?
</blockquote></details>

yes, there can be other errors thrown like in the case of an empty secret identifier, this exception is thrown:
``` InvalidIdentifierRefException("Empty identifier values are not supported")``` 
and we need to propagate this to the user.
___
*[`pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java` line 193 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHrf71-iinUJmW7YV:-NRHeSQ89RZSQdzvLGo3:bakmfad) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/test/java/io/harness/OrchestrationRule.java#L193)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

please use constants 
</blockquote></details>

This is for test and also this string is only used once here.

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 5 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 51 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHHbG34wXKPupZIrDB:-NRIUOm_6y5QHM93wTY4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L51)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Please add some message here to identify exception happened for what use-case. 
</blockquote></details>

Done.
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/utils/FunctorUtils.java` line 18 at r12](https://reviewable.io/reviews/harness/harness-core/45343#-NRIAAcE5GnrhZU8FoHf:-NRIU7M78rg2BoKs5nRz:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/6f99e565b7765534fa66748bd68831d1e18eb854/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/utils/FunctorUtils.java#L18)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This method name (given the class name) sounds too generic. Please CMIIW I think there can be other types of expressions as well right? Or this is the only one possible? Sorry, I don't understand the use-case E2E.
My idea is if there will be more such expressions, then how will we add it here now as the method name used is already very generic.
</blockquote></details>

done

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 5 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java` line 53 at r11](https://reviewable.io/reviews/harness/harness-core/45343#-NRHIKP8DbK4FDg3iQnj:-NRWi8Mq959LQWCBSd3H:b-vehf5x) ([raw file](https://github.com/harness/harness-core/blob/b956c94f8802e1d6c9c24325871425c558d0e713/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/expressions/functors/SecretFunctorWithRbac.java#L53)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

What if there's an NPE somewhere, will it directly propagate to User in this case?
</blockquote></details>

No, we can catch that and throw a specific exception, have made that change.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 2 of 4 files at r8, 9 of 11 files at r9, 1 of 4 files at r12, 3 of 4 files at r13, 2 of 2 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 6 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 4 files at r12, 3 of 4 files at r13, 2 of 2 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45343)* status: 1. unresolved discussions.

**Pull request merged**